File: 1331-rank-transform-of-an-array/1331-rank-transform-of-an-array.cpp
class Solution {
public:
    vector<int> arrayRankTransform(vector<int>& arr) {
        vector<int> sorted = arr;
        // sort
        sort(sorted.begin(), sorted.end());
        
        // Remove duplicates
        sorted.erase(unique(sorted.begin(), sorted.end()), sorted.end());
        
        // Replace each element with its rank
        for (int i = 0; i < arr.size(); i++) {
            // lower_bound returns iterator to first element >= arr[i]
            // Subtract begin to get 0-based index, then +1 for rank
            arr[i] = lower_bound(sorted.begin(), sorted.end(), arr[i]) - sorted.begin() + 1;
        }
        
        return arr;
    }
};

File: 1331-rank-transform-of-an-array/1331-rank-transform-of-an-array.ts
function arrayRankTransform(arr: number[]): number[] {
    // sort
    const sorted = [...arr].sort((a, b) => a - b);
    
    // Remove duplicates
    const unique: number[] = [];
    for (const num of sorted) {
        if (unique.length === 0 || unique[unique.length - 1] !== num) {
            unique.push(num);
        }
    }
    
    // Binary search
    const lowerBound = (nums: number[], target: number): number => {
        let left = 0, right = nums.length;
        while (left < right) {
            const mid = Math.floor((left + right) / 2);
            if (nums[mid] < target) left = mid + 1;
            else right = mid;
        }
        return left;
    };
    
    // Replace with ranks
    const result: number[] = [];
    
    for (const num of arr) {
        const idx = lowerBound(unique, num);
        result.push(idx + 1);
    }
    
    return result;
}